lib/spack/spack/package.py

```diff
@@ -1647,8 +1653,31 @@ def log(self):
             # FIXME : this potentially catches too many things...
             pass
 
+        # Archive the whole stdout + stderr for the package
         install(self.log_path, log_install_path)
+        # Archive the environment used for the build
         install(self.env_path, env_install_path)
+        # Finally, archive files that are specific to each package
+        with working_dir(self.stage.source_path):
+            for glob_expr in self.archive_files:
+                files = glob.glob(glob_expr)
+                for file in files:
+                    try:
+                        target = os.path.join(
+                            spack.store.layout.metadata_path(self.spec),
+                            'archived-files',
+                            file
+                        )
+                        # We must ensure that the directory exists before
+                        # copying a file in
+                        mkdirp(os.path.dirname(target))
+                        install(file, target)
+                    except Exception:
+                        # Here try to be conservative, and avoid discarding
+                        # the whole install procedure because of copying a
+                        # single file failed
+                        tty.warn("FAILED ARCHIVE: {0}".format(file))
```

Comment:
failing to copy a file to a place we control means that something bad happened... maybe this should be a configurable option?

Reply:
I think this should get recorded in the `.spack` directory rather than spit out as a warning.  Maybe keep track of what was missing and spit out one warning and write `.spack/missing.txt` at the end?

Reply:
I am detecting two kinds of errors:
1. Glob expressions that are relative to folders outside the source path
2. Errors when copying a matching file

They'll be recorded in `.spack/archived-files/errors.txt`. In case there's any error, we print a warning and point the user to that file.